docs/proposals/THV-2591-operator-registry-integration-via-gateway-api.md

```diff
@@ -0,0 +1,473 @@
+# Operator Registry Integration via Gateway API
+
+## Problem Statement
+
+ToolHive operators deploy three types of MCP resources in Kubernetes: MCPServer, MCPRemoteProxy, and VirtualMCPServer. Organizations want to expose these servers externally and make them discoverable via the MCP registry, but there's currently no mechanism to:
+
+1. Automatically detect which MCP resources are externally accessible
+2. Populate the registry with external endpoints for these resources
+3. Keep registry entries synchronized with ingress configuration changes
+
+Without this integration, administrators must manually maintain registry entries, leading to configuration drift and stale registry data.
+
+## Goals
+
+- Enable automatic registry population based on Gateway API HTTPRoute resources
+- Provide annotation-based fallback for organizations not using Gateway API
+- Maintain separation between ingress configuration and MCP resource definitions
+- Support all three MCP resource types (MCPServer, MCPRemoteProxy, VirtualMCPServer)
+- Give administrators explicit control over what gets exposed in the registry
+
+## Non-Goals
+
+- Automatic Gateway API resource creation (administrators explicitly create HTTPRoutes)
+- Support for Ingress v1 or other non-Gateway API ingress types
+- Multi-cluster registry aggregation
+
+## Proposed Solution
+
+### Architecture
+
+The operator watches Gateway API resources to discover which MCP servers are externally accessible. Administrators explicitly mark HTTPRoutes for registry export using an annotation. When an annotated HTTPRoute references an MCP resource's Service, the operator:
+
+1. Detects the annotated HTTPRoute and validates it's accepted by the Gateway controller
+2. Extracts the base MCP path from HTTPRoute rules
+3. Traverses to parent Gateway resource(s) to get external hostname/IP
```

Comment:
1. Do we think users will be ok with the Operator being able to read the Gateway API resources and traversing them to build a URL? For MCP related resources its perhaps not a problem, but for others, it seems a bit "grabby"
2. Do we need to traverse to the Gateway resource to get hostname? The hostnames are provided in the `HTTPRoute` right? Perhaps we don't need to go any further to avoid grabbing what we don't need (see point 1)